### PR TITLE
#28813: Moving OpenAI Model paramenters to dotAI App configuration

### LINKED ISSUE #28813: Parametrized OpenAI Models
### Parent Issue

*No response*

### User Story

Currently, the OpenAI models we support, and the parameters for those models \(such as maximum tokens\) are hard-coded in a Java file: https://github.com/dotCMS/core/blob/master/dotCMS/src/main/java/com/dotcms/ai/util/OpenAIModel.java

However, these models and parameters are not static.  For example, OpenAI has added new models which are not in our list, and has already deprecated the gpt-3.5-turbo-16k model \(and will decommission it completely in 3 months\).  But the only way we can change those for dotAI is with a Java code change \(which will then also require backporting to 24.04 LTS\).

We will have a more future-proof solution for this in dotAI phase 2.  However, for now we need to:

1. Not restrict the model name used in our code.
    * Users should be able to type in any model name they want, and we will pass it through to OpenAI.
    * But, we need to be able to detect in code if the OpenAI API returns an error indicating either that the model name is invalid, or that the model is no longer supported.
2.  Move all of the parameters specified in the Java code to configuration parameters in the dotAI App.
    * The code should then pull those parameters from the App.
    * Note that we will need 2 sets of parameters in the dotAI App - one for the text model and one for the image model.

Because of both our inability to support new OpenAI models, and the impending decommissioning of a model we support, this change is a high priority.

### Acceptance Criteria

1. Enable configuration of each of the following models and parameter sets via the dotAI App:
2. When an error is returned from OpenAI indicating an invalid model name was entered, we capture it.
    * Although it would be ideal to differentiate between a model which doesn't exist vs. a model that's no longer supported, I'm not sure the OpenAI API will make that distinction, so this isn't required.
3. When an error is returned from OpenAI indicating that the parameters are wrong \(e.g. maximum tokens exceeded\), we capture it.
4. Any time we capture one of the above types of errors, we do all of the following:
    * Display a popup notification to the user attempting to use dotAI.
        * For the AI blocks, this notification can be displayed in the existing popups or in a separate error popup.
        * For other places where dotAI is used, we display the error in the standard way for that modality.
            * For example, if an error is generated with an API method, we return an error code but do not notify any users.
    * Display a Notification to the admin \(in the Notifications pulldown; same permissions as reindex messages\).
    * Write a message to the log \(ERROR level\).

### Proposed Objective

Reliability

### Proposed Priority

Priority 2 - Important

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

*No response*

### Assumptions & Initiation Needs

*No response*

### Quality Assurance Notes & Workarounds

*No response*

### Sub-Tasks & Estimates

*No response*
\`\``[tasklist]
### Tasks
- [ ] https://github.com/dotCMS/core/issues/29281
- [ ] https://github.com/dotCMS/core/issues/29284
- [ ] https://github.com/dotCMS/core/issues/29587
`\`\`

### LINKED ISSUE #29281: Relocate OpenAI model parameters to dotAI App params
### Parent Issue
https://github.com/dotCMS/core/issues/28813

### User Story
As a Java developer, I want to be able to relocate the hardcoded OpenAI models in our code to the dotAI application, specifically as parameters for each field that conform a model: model names, tokens per minute, API per minute, max number of tokens, and if it's a completion model flag. The changes will include:
- The removal of hardcoded models from the code
- Adding necessary fields to `dotAI.yml` Application descriptor file
- Making all the code adjustments for OpenAI client logic to take the model from this new source instead of code and specify it in the OpenAI request
- Changes to dotAI App UI need to be added in order for the app form validates that the models are valid by looking them up in HTTP request to OpenAI `models` endpoint.

Consider that `Model Names` field is actually a comma delimited list of models.
This will need to happen for both text and image models and changes in the `dotAI.yml` file will result in having UI changes in the Apps portlet that QA will see when testing.

### Acceptance Criteria
- All hardcoded OpenAI models are removed from the code.
- `dotAI.yml` Application descriptor file includes the fields: model names, tokens per minute, API per minute, max number of tokens, and completion model flag.
- The OpenAI client logic retrieves model parameters from `dotAI.yml` instead of the code.
- Adjustments are made for both text and image models.
- Successful end-to-end testing confirming the application uses models defined in `dotAI.yml`.

### dotCMS Version
master

### Proposed Objective
Core Features

### Proposed Priority
Priority 2 - Important

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.
_to_define_

### Assumptions & Initiation Needs
- `dotAI.yml` format and structure are predefined and agreed upon.
- Development environment is set up with access to the necessary OpenAI services.
- All stakeholders are informed about the change and have agreed upon the migration plan.

### Quality Assurance Notes & Workarounds
- Comprehensive testing to ensure no regressions occur due to the removal of hardcoded models.
- Validate the `dotAI.yml` configurations are correctly read and applied in various scenarios.
- Start up a dotCMS instance with a starter that includes recent AI-generation changes to SEO for Blog content type
- Make sure valid license is provided
- Configure dotAI app with a valid API key and a model for text, image and embeddings: `gpt-3.5-turbo-16k`, `dall-e-3` and `text-embedding-ada-002` respectively. Add the following at the `Auto Index Content Config` field:
```
{
    "aiBlog": "blog"
}
```
- Open a `Blog` content, go to the `SEO` tab and verify that AI content is generated by hitting the button for SEO metadata field
- Go to `Dev Tools` -> `dotAI`, at `Manage Embeddings/Indexes` tab, at `Index Name` specify: `aiBlog`, at `Content to Index by Query:` field specify: `+contentType:blog` and build index based on that. Verify that there are index created/updated.

### PR DESCRIPTION
**Pull request by victoralfaro-dotcms**
Removing hardcoded OpenAI models at enum class OpenAIModel. Instead they are now part of the `dotAI.yml` application descriptor so the user can be the one who configures them not only one but multiple models for `text`, `image` and `embeddings`.
The way to specify more than one is to provide a comma delimited list in the new dotAI App params.
Sometimes we accept a model to use in the payload of our AI endpoints, for this matter we will validate that model and if it's invalid we will throw an exception.
When is not present in the payload, then our backend will inject the current model.
Which leads us to the question: How will the current model be determined?
This is part of the work defined for https://github.com/dotCMS/core/issues/29284 (model fallback)

This PR fixes: #28813

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/ai/app/AIModels.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    private AIModels() {
+        // Private constructor to prevent instantiation
+    }
+
+    public static AIModels get() {
+        return INSTANCE.get();
+    }
+
+    void loadModels(final List<AIModel> models, final boolean force) {
+        if (hasLoaded() || force) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/ai/app/AIModels.java:32]: isnt !hasLoaded?

File: dotCMS/src/main/java/com/dotcms/ai/service/OpenAIService.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.dotcms.ai.service;
+
+import com.dotmarketing.util.json.JSONObject;
+
+public interface OpenAIService {
+
+    JSONObject getSupportedModels();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/ai/service/OpenAIService.java:7]: we should stop using JSONObject, use Map instead

Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/ai/service/OpenAIService.java:7]: Add doc

File: dotCMS/src/main/java/com/dotcms/ai/app/AIModels.java (hunk 1 above)
Comment (victoralfaro-dotcms) [reply dotCMS/src/main/java/com/dotcms/ai/app/AIModels.java]: Removed this approach,

File: dotCMS/src/main/java/com/dotcms/ai/service/OpenAIService.java (hunk 2 above)
Comment (victoralfaro-dotcms) [reply dotCMS/src/main/java/com/dotcms/ai/service/OpenAIService.java]: Doc added

File: dotCMS/src/main/java/com/dotcms/ai/model/OpenAIModel.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
+package com.dotcms.ai.model;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.io.Serializable;
+
+public class OpenAIModel implements Serializable {
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/ai/model/OpenAIModel.java:8]: javadoc

File: dotCMS/src/main/java/com/dotcms/ai/model/OpenAIModels.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package com.dotcms.ai.model;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.io.Serializable;
+import java.util.List;
+
+public class OpenAIModels implements Serializable {
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/ai/model/OpenAIModels.java:9]: javadoc

**Comment by dsilvam** [comment]
Please include description for the PR @victoralfaro-dotcms

**Review by jcastro-dotcms — APPROVED** [review]

File: dotcms-integration/src/test/java/com/dotcms/http/CircuitBreakerUrlTest.java [hunk 5]
```diff
@@ -423,4 +440,34 @@ public void testBadRequest_dontThrow() throws Exception {
         assert (cburl.response() >= 400 && cburl.response() <= 499);
     }
 
+    /**
+     * Given an int status code
+     * Then evaluate it does have a SUCCESS http status
+     */
+    @Test
+    public void test_isSuccessStatusCode() {
```
Comment (dsilvam) [inline dotcms-integration/src/test/java/com/dotcms/http/CircuitBreakerUrlTest.java:448]: Why are we changing this test class?

Comment (victoralfaro-dotcms) [reply dotcms-integration/src/test/java/com/dotcms/http/CircuitBreakerUrlTest.java:448]: CircuitBreakerURL logic was included in previous Analytics feature. I need it to relocate it to CircuitBreakerURL in order to be reusable by other classes that need to send HTTP requests. With that change I added an integration test for it.

**Review by dsilvam — APPROVED** [review]

**Pull request merged**